src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,936 @@
+    // the result must not be smaller than:
+    //           1 0 1 0 0 0 0 0
+    // This one satisfies zeros so we can use the logic in the previous case to
+    // obtain our final result, which is:
+    //           1 0 1 0 0 0 1 1
+
+    juint first_violation = count_leading_zeros(zero_violation);
+    // This mask out all bits from the first violation
+    //           1 1 1 1 1 0 0 0
+    U find_mask = ~(std::numeric_limits<U>::max() >> first_violation);
+    //           1 0 0 1 1 1 1 0
+    U either = lo | bits._zeros;
+    // i is the last bit being 0 in either that stands before the first
+    // violation, which is the last set bit of tmp
+    //           0 1 1 0 0 0 0 0
+    U tmp = ~either & find_mask;
+    // i == 2 here, shortcut the calculation instead of explicitly spelling out
+    // i
+    //           0 0 1 0 0 0 0 0
+    U alignment = tmp & (-tmp);
+    // Set the bit at i and unset all the bit after, this is the smallest value
+    // that satisfies bits._zeros. Similar to the above case, this is similar
+    // to aligning lo upto alignment
+    //           1 0 1 0 0 0 0 0
+    U new_lo = (lo & -alignment) + alignment;
+    // Satisfy bits._ones
+    //           1 0 1 0 0 0 1 1
+    new_lo |= bits._ones;
+    assert(lo < new_lo || new_lo == bits._ones, "overflow must return bits._ones");
+    return new_lo;
+  }
+}
+
+// Try to tighten the bound constraints from the known bit information. I.e, we
+// find the smallest value not smaller than lo, as well as the largest value
+// not larger than hi both of which satisfy bits
+// E.g: lo = 0010, hi = 1001
+// zeros = 0011
+// ones  = 0000
+// -> 4-aligned
+//
+//         0    1    2    3    4    5    6    7    8    9    10
+//         0000 0001 0010 0011 0100 0101 0110 0111 1000 1001 1010
+// bits:   ok   .    .    .    ok   .    .    .    ok   .    .
+// bounds:           lo                                 hi
+// adjust:           --------> lo                  hi <---
+template <class U>
+static AdjustResult<RangeInt<U>>
+adjust_bounds_from_bits(const RangeInt<U>& bounds, const KnownBits<U>& bits) {
+  U new_lo = adjust_lo(bounds._lo, bits);
+  if (new_lo < bounds._lo) {
+    // This means we wrapped around, which means no value not less than lo
+    // satisfies bits
+    return AdjustResult<RangeInt<U>>::make_empty();
+  }
+
+  // We need to find the largest value not larger than hi that satisfies bits
+  // One possible method is to do similar to adjust_lo, just with the other
+  // direction
+  // However, we can observe that if v satisfies {bits._zeros, bits._ones},
+  // then ~v would satisfy {bits._ones, bits._zeros}. Combine with the fact
+  // that ~ is a strictly decreasing function, if new_hi is the largest value
+  // not larger than hi that satisfies {bits._zeros, bits._ones}, then ~new_hi
+  // is the smallest value not smaller than ~hi that satisfies
+  // {bits._ones, bits._zeros}
```

Comment:
This is a really nice high level argument. How hard do you think it would be to make it a little more detailed?
It is especially the "strictly decreasing function" argument that might not be very easy for everyone to understand...

Reply:
Done, I have added a small (albeit IMO trivial) proof to this claim.